src/connect/context/ItemsContext.tsx

```diff
@@ -0,0 +1,72 @@
+import { TreeItem, BaseTreeItem, UITreeItem } from '@/connect/components/tree-view-item';
+import React, { useContext, useState } from 'react';
+
+export interface TreeItemContext {
+    items: TreeItem[];
+    uiState: UITreeItemState;
+    baseItems: BaseTreeItem[];
+    virtualItems: BaseTreeItem[];
+    setUIState: React.Dispatch<React.SetStateAction<UITreeItemState>>;
+    setBaseItems: React.Dispatch<React.SetStateAction<BaseTreeItem[]>>;
+    setVirtualItems: React.Dispatch<React.SetStateAction<BaseTreeItem[]>>;
+    setItems: React.Dispatch<React.SetStateAction<BaseTreeItem[]>>;
+}
+
+const defaultTreeItemContextValue: TreeItemContext = {
+    items: [],
+    uiState: {},
+    baseItems: [],
+    virtualItems: [],
+    setItems: () => {},
+    setUIState: () => {},
+    setBaseItems: () => {},
+    setVirtualItems: () => {},
+};
+
+export const ItemsContext = React.createContext<TreeItemContext>(
+    defaultTreeItemContextValue,
+);
+
+export interface ItemsContextProviderProps {
+    children?: React.ReactNode;
+    items?: BaseTreeItem[];
+}
+
+// eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
+export interface UITreeItemState {
+    [key: string]: UITreeItem;
+}
+
+export const ItemsContextProvider: React.FC<ItemsContextProviderProps> = ({
+    children,
+    items: initialItems,
+}) => {
+    const [uiState, setUIState] = useState<UITreeItemState>({});
+    const [virtualItems, setVirtualItems] = useState<Array<BaseTreeItem>>([]);
+    const [baseItems, setBaseItems] = useState<Array<BaseTreeItem>>(initialItems || []);
+
```

Comment:
what are the implications of doing a shallow copy with spread operators here? Perhaps it would be worth breaking out Immer for this

Reply:
I haven't used Immer before, maybe it's a good idea to give it a try, but I think is better to do it in a different PR as an improvement. Let's keep track of it in the backlog 👍🏻